lib/TokenManager.ts

```diff
@@ -294,6 +285,16 @@ export class TokenManager {
     this.renew = renew.bind(this, sdk, tokenMgmtRef, storage);
     this.on = tokenMgmtRef.emitter.on.bind(tokenMgmtRef.emitter);
     this.off = tokenMgmtRef.emitter.off.bind(tokenMgmtRef.emitter);
+    this.hasExpired = hasExpired.bind(this, tokenMgmtRef);
+  
+    const onTokenExpiredHandler = (key) => {
+      if (options.autoRenew) {
+        this.renew(key);
+      } else {
+        this.remove(key);
+      }
+    };
+    this.on('expired', onTokenExpiredHandler);
 
     setExpireEventTimeoutAll(sdk, tokenMgmtRef, storage);
```

Comment:
App might be closed when tokens are still valid storage, then have expired tokens sit in storage on the next startup. Feels like we need a logic to try to renew tokens during app startup.

Reply:
^ great catch

Reply:
This logic does run on startup. It reads existing tokens and sets timeouts for expiration. If a token is already expired, that timeout would be 0 (still async). The renew operation itself is obviously async. So it's not possible to have the tokens validated and renewed **synchronously** at app startup. The app can/should wait for this case though. This is why `tokenManager.hasExpired` is exposed. The `isAuthenticated` method should do something like:

```
let accessToken = await tokenManager.get('accessToken');
if (tokenManager.hasExpired(accessToken)) {
  accessToken = await tokenManager.renew('accessToken');
}
```

The renew operation itself will handle multiple concurrent requests and ensure only a single renew operation is executed.

Reply:
This is why okra-react has isPending - that covers the time between startup and decision about state.  Since we're looking at shifting the state logic to auth-js in 5.0, it might make sense NOT to worry about this status here in 4.0, but I definitely advocate for keeping that idea when we do get there.

Reply:
Internally `isPending` is waiting on the result of `isAuthenticated`. This logic of renewing tokens when asking `isAuthenticated` was happening before implicitly on tokenManager.get(), but it must be done explicitly now. The 
"active" `autoRenew` feature will work in the background to keep tokens refreshed but we cannot rely on it being completed before the first call to `isAuthenticated` or `updateAuthState`. In 5.0 we will handle all this in okta-auth-js, borrowing the implementation from React, but for the 4.0 release, each downstream SDK must handle this.